packages/token-generation-readmodel-checker-verifier/src/utils/utils.ts

```diff
@@ -0,0 +1,1254 @@
+        b && b.length > 0
+          ? PartialTokenStatesPurposeEntry.array().parse(b)
+          : undefined,
+        undefined,
+      ]);
+      return acc;
+    }
+
+    const purposeState = getPurposeStateFromPurposeVersions(c.versions);
+    const lastPurposeVersion = getLastPurposeVersion(c.versions);
+
+    const { status: isPlatformStatesCorrect, data: platformPurposeEntryDiff } =
+      validatePurposePlatformStates({
+        platformPurposeEntry: a,
+        purpose: c,
+        purposeState,
+        lastPurposeVersion,
+      });
+
+    const {
+      status: isTokenGenerationStatesCorrect,
+      data: tokenPurposeEntryDiff,
+    } = validatePurposeTokenGenerationStates({
+      tokenEntries: b,
+      purpose: c,
+      purposeState,
+      lastPurposeVersion,
+    });
+
+    if (!isPlatformStatesCorrect || !isTokenGenerationStatesCorrect) {
+      // eslint-disable-next-line functional/immutable-data
+      acc.push([
+        platformPurposeEntryDiff,
+        tokenPurposeEntryDiff,
+        PartialPurpose.parse(c),
+      ]);
+    }
+
+    return acc;
+  }, []);
+}
+
+function validatePurposePlatformStates({
+  platformPurposeEntry,
+  purpose,
+  purposeState,
+  lastPurposeVersion,
+}: {
+  platformPurposeEntry: PlatformStatesPurposeEntry | undefined;
+  purpose: Purpose;
+  purposeState: ItemState;
+  lastPurposeVersion: PurposeVersion;
+}): {
+  status: boolean;
+  data: PartialPlatformStatesPurposeEntry | undefined;
+} {
+  const isArchived = lastPurposeVersion.state === purposeVersionState.archived;
+  const status = !platformPurposeEntry
+    ? isArchived
+    : !isArchived &&
+      extractIdFromPlatformStatesPK<PurposeId>(platformPurposeEntry.PK).id ===
+        purpose.id &&
+      purposeState === platformPurposeEntry.state &&
+      platformPurposeEntry.purposeConsumerId === purpose.consumerId &&
+      platformPurposeEntry.purposeEserviceId === purpose.eserviceId &&
+      platformPurposeEntry.purposeVersionId === lastPurposeVersion.id;
+
+  return {
+    status,
+    data:
+      !status && platformPurposeEntry
+        ? {
+            PK: platformPurposeEntry.PK,
+            state: platformPurposeEntry.state,
+            purposeConsumerId: platformPurposeEntry.purposeConsumerId,
+            purposeEserviceId: platformPurposeEntry.purposeEserviceId,
+            purposeVersionId: platformPurposeEntry.purposeVersionId,
+          }
+        : undefined,
+  };
+}
+
+function validatePurposeTokenGenerationStates({
+  tokenEntries,
+  purpose,
+  purposeState,
+  lastPurposeVersion,
+}: {
+  tokenEntries: TokenGenerationStatesClientPurposeEntry[] | undefined;
+  purpose: Purpose;
+  purposeState: ItemState;
+  lastPurposeVersion: PurposeVersion;
+}): {
+  status: boolean;
+  data: PartialTokenStatesPurposeEntry[] | undefined;
+} {
+  if (!tokenEntries || tokenEntries.length === 0) {
+    return {
+      status: true,
+      data: undefined,
+    };
+  }
+
+  const foundEntries = tokenEntries.filter(
```

Comment:
I may have misunderstood, but isn't this evaluating as error a valid record in token generation states that has a different purpose than the current one?